source/extensions/filters/network/redis_proxy/command_splitter_impl.cc

```diff
@@ -432,13 +446,98 @@ void SplitKeysSumResultRequest::onChildResponse(Common::Redis::RespValuePtr&& va
+    }
+    transaction.start();
+    // Respond to MULTI locally.
+    localResponse(callbacks, "OK");
+    return nullptr;
+
+  } else if (to_lower_string == "exec" || to_lower_string == "discard") {
+    // Handle the case where we don't have an open transaction.
+    if (transaction.active_ == false) {
+      callbacks.onResponse(Common::Redis::Utility::makeError(
+          fmt::format("{} without MULTI", absl::AsciiStrToUpper(to_lower_string))));
+      return nullptr;
+    }
+
+    // Handle the case where the transaction is empty.
+    if (transaction.key_.empty()) {
+      if (to_lower_string == "exec") {
+        localResponse(callbacks, "(empty array)");
+      } else {
+        localResponse(callbacks, "OK");
+      }
+      transaction.close();
+      return nullptr;
+    }
+
+    // In all other cases we will close the transaction connection after sending the last command.
+    transaction.should_close_ = true;
+  }
+
+  // When we receive the first command with a key we will set this key as our transaction
+  // key, and then send a MULTI command to the node that handles that key.
+  const auto route = router.upstreamPool(transaction.key_);
+  if (transaction.key_.empty()) {
+    transaction.key_ = incoming_request->asArray()[1].asString();
+    Common::Redis::RespValueSharedPtr multi_request =
+        std::make_shared<Common::Redis::Client::MultiRequest>();
+    if (route) {
+      makeSingleServerRequest(route, "MULTI", transaction.key_, multi_request, null_pool_callbacks,
+                              callbacks.transaction());
+    }
+  }
+
+  std::unique_ptr<TransactionRequest> request_ptr{
+      new TransactionRequest(callbacks, command_stats, time_source, delay_command_latency)};
+  if (route) {
+    Common::Redis::RespValueSharedPtr base_request = std::move(incoming_request);
+    request_ptr->handle_ =
+        makeSingleServerRequest(route, base_request->asArray()[0].asString(), transaction.key_,
+                                base_request, *request_ptr, callbacks.transaction());
+  }
+
+  if (!request_ptr->handle_) {
+    command_stats.error_.inc();
+    callbacks.onResponse(Common::Redis::Utility::makeError(Response::get().NoUpstreamHost));
+    return nullptr;
+  }
+
+  return request_ptr;
+}
+
 InstanceImpl::InstanceImpl(RouterPtr&& router, Stats::Scope& scope, const std::string& stat_prefix,
                            TimeSource& time_source, bool latency_in_micros,
                            Common::Redis::FaultManagerPtr&& fault_manager)
     : router_(std::move(router)), simple_command_handler_(*router_),
       eval_command_handler_(*router_), mget_handler_(*router_), mset_handler_(*router_),
       split_keys_sum_result_handler_(*router_),
-      stats_{ALL_COMMAND_SPLITTER_STATS(POOL_COUNTER_PREFIX(scope, stat_prefix + "splitter."))},
+      transaction_handler_(*router_), stats_{ALL_COMMAND_SPLITTER_STATS(
+                                          POOL_COUNTER_PREFIX(scope, stat_prefix + "splitter."))},
```

Comment:
nit: might be better to add `transaction_handler_(*router_),` in a separate line

Reply:
Fixed.